Fixes #27 - allow loading of external fact hashes

logicminds opened the pull request:
* previously there was no way to supply facterdb with
    external facts that contain sensitive information.  This
    fixes that by allowing the user to set an environment variable
    that take precedence over the facts from this gem.

mcanevet commented:
LGTM. @DavidS @rodjek what do you think about this PR?

rodjek commented on Gemfile:
```diff
@@ -7,3 +7,7 @@ if facterversion = ENV['FACTER_GEM_VERSION']
 end
+
+group :dev do
+  gem 'pry'
```
Can you add this to the gemspec as a development dependency instead of here?

logicminds replied:
added to gemspec

DavidS replied:
then you can remove it here?

rodjek commented on lib/facterdb.rb:
```diff
@@ -4,7 +4,35 @@
+  def self.external_fact_files(fact_paths = ENV['FACTERDB_SEARCH_PATH'])
+    fact_paths ||= ''
+    return [] if fact_paths.empty?
+    paths = fact_paths.split(',').map do |fact_path|
```
Rather than splitting by `,`, can we change this to `File::PATH_SEPARATOR` which will be a bit more idiomatic to users (`:` on Linux/BSD/etc, `;` on Windows)

rodjek commented on lib/facterdb.rb:
```diff
@@ -4,7 +4,35 @@
+      end
+      File.join(fact_path.strip, '**', '*.facts')
+    end.compact
+    Dir.glob(paths)
```
`Dir.glob` behaves badly when passed Windows style backslash delimited paths, so I'd recommend adding the following line to the above map before the final `File.join` to convert the backslashes.

```ruby
fact_path = fact_path.gsub(File::ALT_SEPARATOR, File::SEPARATOR) if File::ALT_SEPARATOR
```

logicminds replied:
These are great suggestions.  Fixed in latest push.

logicminds commented:
This is ready to go after adding the suggested fixes. @rodjek

DavidS commented:
I like the idea, and given that there is already https://github.com/mcanevet/rspec-puppet-facts#set-global-custom-facts there is surely a need for it, too. In the light of #61 , I wonder if it wouldn't make sense to extend this even further and make the first/default entry of `FACTERDB_SEARCH_PATH` a special value, like `BUILTIN`, so that folks can override the DB contents completely, if they want?

logicminds commented:
@DavidS we can do something like this:

```
# @return [Array[String]] -  list of all files found in the default facterdb facts path
  def self.default_fact_files
    proj_root = File.join(File.dirname(File.dirname(__FILE__)))
    facts_dir = ENV['FACTERDB_DEFAULT_SEARCH_PATH'] || File.expand_path(File.join(proj_root, 'facts'))
    Dir.glob(File.join(facts_dir, "**", '*.facts'))
  end
```

logicminds commented:
Also now that I look at this.  I am going to change `FACTERDB_SEARCH_PATH` to `FACTERDB_SEARCH_PATHS` since we can supply multiple paths.

DavidS commented:
`*_PATH` sounds fine to me.

Re the default search path, I've spent some more time on that, and I think we should optimize strongly for the default case of folks using the built-in facts. What about `return [] if ENV['FACTERDB_SKIP_DEFAULTDB'] == 'yes'`, adding "Set FACTERDB_SKIP_DEFAULTDB=yes to instruct facterdb to look at its built-in facts, if you want to completely replace which facts are used." to the README.

logicminds commented:
I have added @DavidS suggestions and updated the readme to include this additional info on how to skip the default database.  I think this feature will be a big hit.

DavidS commented on Gemfile:
```diff
@@ -6,4 +6,4 @@ if facterversion = ENV['FACTER_GEM_VERSION']
 else
   gem 'facter', :require => false
-end
+end
```
:-(

logicminds replied:
I didn't want to get rubocop involved as it would have changed a whole lot of code.  But now that this is merged, let the games begin. Also we need to bump the version.

DavidS replied:
rubocop++, and I would suggest the [PDK template](https://github.com/puppetlabs/pdk/blob/master/.rubocop.yml), obviously. But final say is with @mcanevet and @raphink . 

the version gets bumped during release-prep in accordance with the changes in the release.

mcanevet replied:
@DavidS I didn't have time to check PDK yet but feel free to add this in this module. Ping me onde it's OK for release.
